plugins/artifacts/artifacts.go

```diff
@@ -0,0 +1,128 @@
+//go:generate go-composite-schema --unexported artifacts payload-schema.yml generated_payloadschema.go
+
+// Package artifacts is responsible for uploading artifacts after builds
+package artifacts
+
+import (
+	"mime"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+type pluginProvider struct {
+	extpoints.PluginProviderBase
+}
+
+func (pluginProvider) NewPlugin(extpoints.PluginOptions) (plugins.Plugin, error) {
+	return plugin{}, nil
+}
+
+type plugin struct {
+	plugins.PluginBase
+}
+
+func (plugin) PayloadSchema() (runtime.CompositeSchema, error) {
+	return payloadSchema, nil
+}
+
+func (plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	if options.Payload == nil {
+		return plugins.TaskPluginBase{}, nil
+	}
+	return &taskPlugin{
+		TaskPluginBase: plugins.TaskPluginBase{},
+		payload:        *(options.Payload.(*payload)),
+	}, nil
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	context *runtime.TaskContext
+	payload payload
+}
+
+func (tp *taskPlugin) Prepare(context *runtime.TaskContext) error {
+	tp.context = context
+	return nil
+}
+
+func (tp *taskPlugin) Stopped(result engines.ResultSet) (bool, error) {
+	var err error
+	for _, artifact := range tp.payload {
+		// If expires is set to this time it's either the default value or has been set to an invalid time anyway
+		if time.Time(artifact.Expires).IsZero() {
+			artifact.Expires = tp.context.TaskInfo.Expires
+		}
+		switch artifact.Type {
+		case "directory":
+			err = result.ExtractFolder(artifact.Path, tp.createUploadHandler(artifact.Name, artifact.Path, artifact.Expires))
+			if err != nil {
+				if tp.errorHandled(artifact.Name, artifact.Expires, err) {
+					continue
+				}
+				return false, err
+			}
+		case "file":
+			fileReader, err := result.ExtractFile(artifact.Path)
+			if err != nil {
+				if tp.errorHandled(artifact.Name, artifact.Expires, err) {
+					continue
+				}
+				return false, err
+			}
+			err = tp.attemptUpload(fileReader, artifact.Path, artifact.Name, artifact.Expires)
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+	return true, nil
+}
+
+func (tp taskPlugin) errorHandled(name string, expires tcclient.Time, err error) bool {
+	if err == engines.ErrFeatureNotSupported || err == engines.ErrResourceNotFound ||
+		err == engines.ErrNonFatalInternalError || err == engines.ErrHandlerInterrupt ||
+		reflect.TypeOf(err).String() == "engines.MalformedPayloadError" {
```

Comment:
Avoid using reflect...

Just make different if statement on the form:
```go
if e, ok := err.(*engines.MalformedPayloadError); ok {
  // Now is e is a MalformedPayloadError type
}
```

Don't for forget the `ok` part, otherwise it'll panic on failure.